fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/ResolverValidator.java

```diff
@@ -0,0 +1,301 @@
+     */
+    public static RecordCursor<ValidatedEntry> validate(@Nonnull final LocatableResolver resolver,
+                                                        @Nonnull final FDBRecordContext context,
+                                                        @Nullable final byte[] continuation,
+                                                        final int reverseLookupPipelineSize,
+                                                        final boolean repairMissingEntries,
+                                                        @Nonnull final ScanProperties scanProperties) {
+        return resolver.scan(context, continuation, scanProperties)
+                .mapPipelined(keyValue -> {
+                    // The reverse directory cache has an entry in the directory layer that it does not,
+                    // itself maintain a reverse entry for (it probably SHOULD, but we now have caches that
+                    // were created that way and have to accept it.
+                    if (keyValue.getKey().equals(FDBReverseDirectoryCache.REVERSE_DIRECTORY_CACHE_ENTRY)) {
+                        return CompletableFuture.completedFuture(new ValidatedEntry(ValidationResult.OK, keyValue));
+                    }
+
+                    return resolver.reverseLookup(context.getTimer(), keyValue.getValue().getValue())
+                            .handle((reverseKey, exception) -> {
+                                if (exception != null) {
+                                    if (isEntryMissing(exception)) {
+                                        return new ValidatedEntry(ValidationResult.REVERSE_ENTRY_MISSING, keyValue);
+                                    }
+
+                                    throw new RecordCoreException("Error reading reverse directory entry", exception)
+                                            .addLogInfo(LogMessageKeys.RESOLVER, resolver)
+                                            .addLogInfo(LogMessageKeys.RESOLVER_KEY, keyValue.getKey())
+                                            .addLogInfo(LogMessageKeys.RESOLVER_VALUE, keyValue.getValue().getValue())
+                                            .addLogInfo(LogMessageKeys.RESOLVER_METADATA, ByteArrayUtil2.loggable(keyValue.getValue().getMetadata()));
+                                }
+
+                                if (!reverseKey.equals(keyValue.getKey())) {
+                                    return new ValidatedEntry(ValidationResult.REVERSE_ENTRY_MISMATCH, keyValue, reverseKey);
+                                }
+
+                                return new ValidatedEntry(ValidationResult.OK, keyValue);
+                            }).thenCompose(validatedEntry -> {
+                                if (repairMissingEntries
+                                        && validatedEntry.getValidationResult() == ValidationResult.REVERSE_ENTRY_MISSING) {
```

Comment:
In theory, this could also repair `REVERSE_ENTRY_MISMATCH` errors by replacing the wrong key with the right key in the reverse mapping. Is the idea that that represents something "worse" than just a missing entry (which, for example, could be caused by something like resolving something in a directory layer prior to the reverse directory layer having been added)? (The most plausible cause for a `REVERSE_ENTRY_MISMATCH`, off the top of my head, would be something like a concurrency bug that would accidentally result in two directory layer entries being mapped to the same value, which seems bad.)